cli/src/lockfiles/mod.rs

```diff
@@ -22,6 +22,13 @@ pub type ParseResult = anyhow::Result<Vec<PackageDescriptor>>;
 
 pub trait Parseable {
     fn new(filename: &Path) -> Result<Self, io::Error>
+    where
+        Self: Sized,
+    {
+        Ok(Self::from_string(std::fs::read_to_string(filename)?))
+    }
+
+    fn from_string(text: String) -> Self
```

Comment:
With a `parse()` and a `from_string()` method, we're kind of dancing around the [`FromStr` trait](https://doc.rust-lang.org/std/str/trait.FromStr.html). We could almost implement this as:

```
pub trait Parseable: FromStr<Err=anyhow::Error> + Into<Vec<PackageDescriptor>> { }
```

Reply:
`FromStr` is fallible though, while this `from_string` method is not. The ownership is also significant here since we otherwise would need to clone it internally.

Reply:
Right. This would only work in combination with your other comment about making the parsers take `&str`. And then `T::from_string(foo).parse()` becomes `foo.parse::<T>()`

Reply:
I feel like the `FromStr` trait often gets in the way more than it actually benefits people. Especially because trait discovery is often harder than finding `::new` or similar methods.